prog/dftb+/lib_dftbplus/mainapi.F90

```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+    integer,  allocatable, intent(in) :: species0(:)
+    character(mc),  allocatable, intent(in) :: speciesName(:)
+    !> Data type for atomic orbitals
+    type(TOrbitals),       intent(in) :: orb
+    !> Number of electrons
+    real(dp),              intent(in) :: nElectrons(:)
+    !> Orbital equivalence relations
+    integer,  allocatable, intent(in) :: iEqOrbitals(:,:,:)
+    !> nr. of inequivalent orbitals
+    integer,               intent(in) :: nIneqOrb
+    !> nr. of elements to go through the mixer - may contain reduced orbitals and also orbital blocks                
+    !> (if tDFTBU or onsite corrections)                                           
+    integer, intent(in) :: nMixElements
+
+    !> Initial spins
+    real(dp), allocatable, intent(in), optional :: initialSpins(:,:)
+    !> Set of atom-resolved atomic charges 
+    real(dp), allocatable, intent(in), optional :: initialCharges(:)
+ 
+    !> reference neutral atomic occupations
+    real(dp), allocatable, intent(inout) :: q0(:, :, :)
+    !> shell resolved neutral reference
+    real(dp), allocatable, intent(inout) :: qShell0(:,:)
+    !> input charges (for potentials)
+    real(dp), allocatable, intent(inout) :: qInput(:, :, :)
+    !> output charges
+    real(dp), allocatable, intent(inout) :: qOutput(:, :, :)
+    !> input charges packed into unique equivalence elements
+    real(dp), allocatable, intent(inout) :: qInpRed(:)
+    !> output charges packed into unique equivalence elements
+    real(dp), allocatable, intent(inout) :: qOutRed(:)
+    !> charge differences packed into unique equivalence elements
+    real(dp), allocatable, intent(inout) :: qDiffRed(:)
+
+    !> TODO(Alex) Not global in initprogram. Currently hard-coded
+    !> Is check-sum for charges read externally to be used?
+    logical  :: tSkipChrgChecksum = .false.
+
+    !Local variables 
+    integer        :: iAt,iSp,iSh,ii,jj,   i,j
+
+    if (.not. tMulliken .or. .not. tSccCalc) then
+       return
+    endif
+
+    !Support to be added 
+    if(tReadChrg) then
+       call error("Reading charges from file is not supported by DFTB+ API")
+    endif
+    
+    if(tMixBlockCharges) then
+       call error("Mixing block charges is not supported by DFTB+ API")
+    endif
+
+    if(tRangeSep) then
+       call error("Range-separated charges is not supported by DFTB+ API")
+    endif
+
+    if(tCustomOccAtoms) then
+       call error("Custom occupation is not supported by DFTB+ API")
+    endif
+
+    @:ASSERT(size(species0) == nAtom) 
+    !call error("Number of atoms in species array differ from current state of DFTB+")
+
+    ! Initialise reference neutral atoms 
+    if(.not. allocated(q0)) allocate(q0(orb%mOrb, nAtom, nSpin))
+    q0(:,:,:) = 0.0_dp
+    !referenceN0 consistent as long as speciesName remains constant
+    call initQFromShellChrg(q0, referenceN0, species0, orb)
+
+    !TODO(Alex) This does not appear to get used in any instance in DFTB+. Remove?
```

Comment:
Is qShell0 required?

Reply:
Probably not. Nevertheless, we must this factorize out in `initprogram.F90` to avoid code duplication.